pwndbg/gdblib/dynamic.py

```diff
@@ -0,0 +1,801 @@
+        return LinkMapEntry(r_map)
+
+
+# @pwndbg.lib.cache.cache_until("link_map")
+def link_map():
+    """
+    Iterator over all the entries in the link map.
+    """
+    head = link_map_head()
+    while head is not None:
+        yield head
+        head = head.next()
+
+
+class LinkMapEntry:
+    """
+    An entry in the link map.
+    """
+
+    def __init__(self, address):
+        self.link_map = CStruct.link_map()
+        self.link_map_address = address
+
+    def name(self):
+        """
+        The name of the binary image this entry describes.
+        """
+        ptr = self.link_map.read(self.link_map_address, "l_name")
+        return pwndbg.gdblib.memory.string(ptr)
+
+    def dynamic(self):
+        """
+        The pointer to the memory mapped dynamic segment of the binary image.
+        """
+        return self.link_map.read(self.link_map_address, "l_ld")
+
+    def load_bias(self):
+        """
+        The difference between the addresses in the data structures of the
+        binary image and the actual location of the data being pointed to by
+        them in the address space of the inferior. This number will never be
+        negative.
+
+        Aditionally, for DYN images, such as PIE executables and shared
+        libraries, this value is the same as the base load address of the image.
+
+        The term "load bias" comes from the ELF binary format loading procedure
+        in the Linux Kernel.
+        """
+        return self.link_map.read(self.link_map_address, "l_addr")
+
+    def next(self):
+        """
+        The next entry in the chain, if any.
+        """
+        ptr = self.link_map.read(self.link_map_address, "l_next")
+        if ptr == 0:
+            return None
+        else:
+            return LinkMapEntry(ptr)
+
+    def prev(self):
+        """
+        The previous entry in the chain, if any.
+        """
+        ptr = self.link_map.read(self.link_map_address, "l_prev")
+        if ptr == 0:
+            return None
+        else:
+            return LinkMapEntry(ptr)
+
+    def __repr__(self):
+        return f"<{self.__class__.__name__} node={self.link_map_address:#x} name={self.name()} load_bias={self.load_bias():#x} dynamic={self.dynamic():#x}>"
+
+
+# Normally, only one entry for each tag is allowed to be present in the dynamic
+# array for us to consider the dynamic array to be well-formed. Tags in this
+# set are allowed to appear multiple times.
+DYNAMIC_SECTION_ALLOW_MULTIPLE = set([elf.DT_NEEDED])
+
+# The DynamicSegment class expects some tags to always be present to function
+# correctly. In this set we list them explicitly. Code in that class is allowed
+# to presume these tags are always present after __init__.
+DYNAMIC_SECTION_REQUIRED_TAGS = set(
+    [
+        elf.DT_STRTAB,
+        elf.DT_STRSZ,
+        elf.DT_SYMTAB,
+        elf.DT_SYMENT,
+    ]
+)
+
+
+class DynamicSegment:
+    """
+    Parser for the DYNAMIC segment present in a binary image.
```

Comment:
Can't we leverage sth like https://github.com/eliben/pyelftools for such parsing? 

I am a bit worried that we may end up having all this outdated & not supporting new use cases very quickly :/

Reply:
Well, I don't think we can, I considered using it at first, but ended up deciding against it for two main reasons. 

First, `elftools` is fairly set on the case where the entire ELF file is present, and its dynamic segment parsing logic is happy to rely on things it gathered from the ELF header, simply because it already assumes it will be there. While it is technically true that the entire ELF should be present for us to use, too, we don't strictly need it, and having to find it from the pointer to the dynamic segment that we do have would most likely bring back the issues with reliability that I was concerned with. Plus, doing it directly means we can tap into `pwndbg`'s already existing functionality for handling memory through GDB.

Second, most of these structures - including all of the ones we are parsing - are effectively fossilized, from what I can tell. The things we care about have mostly stayed the same for [the past 24 years or so](https://www.sco.com/developers/gabi/latest/revision.html). Though, yes, from what it seems following SVR4 is more of a convention than a rule with GNU, the last time they broke with it in a way that could become relevant to us was [around 2006](https://sourceware.org/legacy-ml/binutils/2006-10/msg00377.html), when they introduced `GNU_HASH`. So I honestly don't think it becoming outdated should be that big of a concern?

Reply:
:+1:, you are right; its all just lots of code, so I was wondering if we can make it less :P

Reply:
Yes, I'm not really comfortable with this much code myself, but I don't really think we have much in the way of good alternatives, at least none that I could think of